### ISSUE TITLE
Add color tokens to tailwind config file

### ISSUE THREAD

**Issue by huyenltnguyen** [freeCodeCamp/freeCodeCamp#42186]
We are in the progress of building a component library, and [TailwindCSS](https://tailwindcss.com/docs) is what we use for component theming.

The config file has been created, and further configurations will be added in the future. For now, we need to add color tokens to the file, in the `colors` object. 

Link to the config file: https://github.com/freeCodeCamp/freeCodeCamp/blob/main/tools/ui-components/tailwind.config.js#L5

Colors to be added:
- Layout colors: https://design-style-guide.freecodecamp.org/#layout-colors
- Accent colors: https://design-style-guide.freecodecamp.org/#accent-colors
- Any colors in here that the design style guide does not have: https://github.com/freeCodeCamp/freeCodeCamp/blob/main/client/src/components/layouts/variables.css

Convention:
- Token names are in camelCase
- Tokens for dark theme colors are suffixed with the word `dark`. Tokens for light theme can stay as is. For example: `primaryColor` and `primaryColorDark`, `secondaryColor` and `secondaryColorDark`, etc.
- Tokens are sorted alphabetically

**Comment by huyenltnguyen** [comment]
@ahmadabdolsaheb I'm using the color palette in the style guide as the starting point, but do you think we need other colors in /client's [variables.css](https://github.com/freeCodeCamp/freeCodeCamp/blob/main/client/src/components/layouts/variables.css) as well?

**Comment by ahmadabdolsaheb** [comment]
Yes we should add variables.css in addition to the design style guide colors. The whole dark mode, layout, and accent colors might need another review, but that will come later.

**Comment by huyenltnguyen** [comment]
Thanks @ahmadabdolsaheb! 

I have updated the issue description to lay out the requirements. I might need to discuss with you on the token naming (I'm considering using `foreground` and `background`), but that is just a minor thing that came into my mind. We can leave it for later since the color palette will be reviewed.

**Comment by huyenltnguyen** [comment]
I'm going to open this issue up for first-time contributor. If you are interested in helping us with this issue, please add a comment here to let us know. 

Please also make sure you read [our guidelines for contributing](https://contribute.freecodecamp.org/#/), we prioritize contributors following the instructions in our guides. Join us in our [chat room](https://chat.freecodecamp.org/channel/contributors) or [the forum](https://forum.freecodecamp.org/c/contributors/3) if you need help contributing, our moderators will guide you through this.

Sometimes we may get more than one pull requests. We typically accept the most quality contribution followed by the one that is made first.

Happy contributing.

**Comment by ChrisPrzR** [comment]
I'd like to work on this.

**Comment by ChrisPrzR** [comment]
There is a "darkviolet" color in the variables.css file.

```--purple-mid: darkviolet;```

I could not find a reference for it in the document, is there any specific color I should use?

**Comment by ahmadabdolsaheb** [comment]
@huyenltnguyen I know there are different ways to setup the night mode. 

How about adding a main.css to our tailwind setup and including all of our variables there.
Then we could set tailwind config colors to CSS variables.

Do you see any pros or cons for this setup?
We could build on @ChrisPrzR's pr if necessary.

**Comment by huyenltnguyen** [comment]
Yeah I was looking for a way to keep the colors in a variable CSS file instead of porting them to the config file as well. [The article](https://levelup.gitconnected.com/tailwindcss-with-css-variables-513abe2e9a5) I read suggested the same approach as the one you shared.

With that approach, we would need to maintain a file with CSS variables and map the variables to a corresponding color property in the tailwind config. That extra mapping step makes me hesitant to keep a CSS file, and I wasn't sure how complex the palette would eventually be, so I opted for a simpler approach.

However, I guess we could consider having two layers for our colors (this is what I said I would discuss with you more): 
- One is for color names, which describes what the colors are. For example: lightGreen, darkGreen, gray05, gray90, etc.
- One is for semantic names, which describes what the colors are used for. For example: primaryText, primaryTextDark, secondaryText, secondaryTextDark, etc.

And to translate to code:

The first layer is in a CSS file

```css
/* colors.css */
:root {
  --gray00: #ffffff;
  --gray90: #0a0a23;
}
```

The second layer is in the config file, whose tokens will then be used in the styling of our components

```js
// tailwind.config.js
theme: {
  colors: {      
    primaryText: 'var(--gray90)',
    primaryTextDark: 'var(--gray00)',
  }
}
```

**Comment by ahmadabdolsaheb** [comment]
I think that is a simple approach. How would the component color change from primaryText to primaryTextDark? 

Do we need to add that logic per component based on the theme props it receives?

**Comment by huyenltnguyen** [comment]
To be honest, I haven't put much thoughts into the dark theme support. I have added an item for this in the main issue, but the traditional `.dark {}` way is what I have at the moment. However, I'm afraid that it might not work well with the CSS purging that Tailwind offers.

Happy to hear you thoughts on this.

**Comment by huyenltnguyen** [comment]
I skimmed the Tailwind doc yesterday and found that it provides a `darkMode` option in the config. I think I would try out and see if I can use the tokens I proposed this it. 

Here is the doc: https://tailwindcss.com/docs/dark-mode#toggling-dark-mode-manually

**Comment by Nirajn2311** [comment]
For step 2, I found a blog where someone changed that default prefix(not sure how effective it will be). Here's the article - https://logaretm.com/blog/2020-02-08-tailwind-dark-mode/

**Comment by Nrestrepo05** [comment]
Hey! I'd like to work on this issue, so now I'm working on a new repo trying to develop the dark mode in a good way. When I finish it, I will share it with you!

**Comment by Nrestrepo05** [comment]
I already finished the repo, here is the link:

https://github.com/Nrestrepo05/darkmode

I think we can still use colors vars and a class on the body tag as we have already been doing. As you can see in my repo I tried to keep doing what is already implemented and it works perfectly.

Hope to know what you think!

**Comment by huyenltnguyen** [comment]
@Nirajn2311 Thanks for the link! I'll take a look. It's just that I have not written code with atomic CSS for a long time, and I never knew I would need to relearn to get back to this (as in I don't know what the right coding pattern should be now) 😅 

---

@Nrestrepo05 First of all, welcome to freeCodeCamp and big thank for the demo repo! I am also in favor of following the current implementation. I personally find it easier to maintain that way (though that could be because I'm so used to this implementation approach); but it will help reduce a lot of configurations and changes we need to make.

The issue's status is "discussing" currently as we have not finalized path to proceed. I'm more than happy to pass it to you once we have things ironed out, so please be patient with us 😄 

In the meantime, please check out other issues with "first timers Only" or "help wanted" label, and come say hi in our [contributors chat room](https://chat.freecodecamp.org/channel/contributors).

**Comment by huyenltnguyen** [comment]
@ahmadabdolsaheb hope you could weigh in on this. To sum up the entire discussion, there are two approaches we can take regarding dark theme support:
- The currently approach in /learn in which we use light theme styles as the default, then override them with the `.night` selector for dark mode.
  - Pros: 
    - Consistent with the implementation of /learn, thus easier to get familiar
    - Less configurations/changes needed
  - Cons:
    - This can be considered anti-pattern as Tailwind is a utility-first library
- The [Tailwind approach](https://tailwindcss.com/docs/dark-mode#toggling-dark-mode-manually) in which we split the styles smaller into atomic CSS and assign their class names to our components (also see https://github.com/freeCodeCamp/freeCodeCamp/issues/42186#issuecomment-856726960)
  - <img width="834" alt="Screen Shot 2021-06-13 at 01 24 13" src="https://user-images.githubusercontent.com/25715018/121785912-d5834a80-cbe6-11eb-9ca2-f21e2c8dc171.png">
  - Pros: 
    - Get the benefits of the library by following its pattern (I'm not sure 100% what the benefits are except for the [CSS purging](https://tailwindcss.com/docs/optimizing-for-production#removing-unused-css))
  - Cons:
    - The introduction of the `dark` class as Tailwind uses it as the theme identifier
    - The HTML may be bloated with all sort of class names. For example: `bg-colorA`, `focus:bg-colorB`, `hover:bg-colorC`, `dark:bg-colorD`, `dark:focus:bg-colorE`,... )

---

I think approach (1) is more viable. This is not saying that we shouldn't use Tailwind classes at all, as I reckon it will be very useful and convenient when it comes to handling layout. But for dark theme styling specifically, I think we are better off with our current way of writing CSS.

**Comment by ahmadabdolsaheb** [comment]
@huyenltnguyen, thank you for the extensive research.

A little bit of history: we use to set the "night" css class in the html and style elements by adding .night to the night mode selectors. We ended up with a lot of styles to maintain. Then, we moved on to contrast based theming and simplified our sheets a bit by using css variables.

Although approach 1 is not tailwind's current official method to implement dark mode, it is well documented in the community ex. https://jeffjadulco.com/blog/dark-mode-react-tailwinds

I think we could use the current convention and keep things dry and straightforward. We could add class based styles later on if we wanted more control.

**Comment by Nrestrepo05** [comment]
sorry @ahmadabdolsaheb I didn't understand so well. Do you want to keep doing things like they are already (css variables) or you want to use the tailwind convention?

**Comment by Nrestrepo05** [comment]
Thank you @ahmadabdolsaheb, do you think there something that I can do on this issue?

**Comment by ahmadabdolsaheb** [comment]
@Nrestrepo05, I am not sure

**Comment by ahmadabdolsaheb** [comment]
Since our colors have evolved since we published the design style guide, there are a couple of notes to take into account when approaching the colors config, naming, etc. in developing the component library:

- there are to main types of colors: the colors that switch during the theme change  (background, foreground) and colors that do not (cta yellow).
- colors that change have less saturation, and are more appropriate for background and foreground.
- all background and foreground colors could be produced using scss mix or css color-mix functions

// colors that change with theme change
Here is a sample of how we could name our colors moving forward (just a suggestion):
gray-foreground-primary
gray-background-primary
red
red-foreground-primary
red-background-primary
blue
capri
gold

Note: The colors that don't change should maintain a good contrast regardless of the theme.
Note: We should be able to programmatically derive all colors using a couple of scsss functions, but I have not looked into it yet. ( might be an overkill )

**Comment by huyenltnguyen** [comment]
Do you see `gray-foreground-primary`, `gray-background-primary`,... as names for color variables (those in `colors.css`), or for the semantic colors (those in the config file)?

My take on this is that, semantic colors won't have anything refer to the actual color they use. An example can be the Alert component. [Bootstrap](https://getbootstrap.com/docs/5.0/components/alerts/) and [Material UI](https://material-ui.com/components/alert/) calls the variants `success`, `info`, `warning`, ... instead of `green`, `blue`, `yellow`.

So I guess `gray-foreground-primary` would be a color in the variable file, and it is mapped to, say, `default-foreground-primary` in the config file. Then `red-foreground-primary` would be something like `danger-foreground-primary`, etc.

With the colors being added to `colors.css`, our next step is to map them to semantic tokens in the config file. Since the naming of colors in both layers will need to be adjusted again once the design style guide gets new updates, are you okay with me coming up with some temporary tokens for now? 

Though I think I should do it incrementally as I build the components, rather than assuming what each color means right now and add their tokens in bulk.

**Comment by ahmadabdolsaheb** [comment]
Yes, your plan makes more sense to me.
